llarp/handlers/tun.cpp

```diff
@@ -585,80 +584,120 @@ namespace llarp
     bool
     TunEndpoint::SetupTun()
     {
+      lazy_vpn vpn;
+      huint32_t ip;
       auto loop = EndpointNetLoop();
-      if(!llarp_ev_add_tun(loop.get(), &tunif))
-      {
-        llarp::LogError(Name(),
-                        " failed to set up tun interface: ", tunif.ifaddr,
-                        " on ", tunif.ifname);
-        return false;
-      }
-
-      struct addrinfo hint, *res = nullptr;
-      int ret;
-
-      memset(&hint, 0, sizeof hint);
-
-      hint.ai_family = PF_UNSPEC;
-      hint.ai_flags  = AI_NUMERICHOST;
-
-      ret = getaddrinfo(tunif.ifaddr, nullptr, &hint, &res);
-      if(ret)
-      {
-        llarp::LogError(Name(),
-                        " failed to set up tun interface, cant determine "
-                        "family from ",
-                        tunif.ifaddr);
-        return false;
-      }
-
-      /*
-      // output is in network byte order
-      unsigned char buf[sizeof(struct in6_addr)];
-      int s = inet_pton(res->ai_family, tunif.ifaddr, buf);
-      if (s <= 0)
-      {
-        llarp::LogError(Name(), " failed to set up tun interface, cant parse
-      ", tunif.ifaddr); return false;
-      }
-      */
-      if(res->ai_family == AF_INET6)
+      if(tunif == nullptr)
       {
-        m_UseV6 = true;
-      }
+        llarp::LogInfo(Name(), " waiting for vpn to start");
+        vpn   = m_LazyVPNPromise.get_future().get();
+        vpnif = vpn.io;
+        if(vpnif == nullptr)
+        {
+          llarp::LogError(Name(), " failed to recieve vpn interface");
+          return false;
+        }
+        llarp::LogInfo(Name(), " got vpn interface");
+        auto self = shared_from_this();
```

Comment:
Why `shared_from_this()` instead of just capturing `this`?

Reply:
would using `this` be safe?

Reply:
I didn't dig too deep there initially, but it looks like `shared_from_this()` has the effect of tying the lifetime of `this` to the `uv_check_t` that gets injected into `libuv`.

Reply:
yes that is why i did it